### PR TITLE
More correct implementation of SE-0110

### PR DESCRIPTION
**Pull request by slavapestov**
Related radars:

- rdar://27383557
- rdar://27857015
- rdar://28621719
- rdar://29409725

### CODE REVIEW

File: test/ClangImporter/foreign_errors.swift [hunk 1]
```diff
@@ -14,7 +14,7 @@ func test0() {
 func testAndReturnError() throws {
   try ErrorProne.fail()
   try ErrorProne.go()
-  try ErrorProne.tryAndReturnError() // collides with 'try' keyword
+  try ErrorProne.tryAndReturnError(()) // collides with 'try' keyword
```
Comment (jrose-apple) [inline test/ClangImporter/foreign_errors.swift]: Ooh, this is actually a change. We shouldn't be adding a dummy argument here.

**Comment by jrose-apple** [comment]
It's worth checking what things we publicly promised would still work after [SE-0110](https://github.com/apple/swift-evolution/blob/master/proposals/0110-distingish-single-tuple-arg.md) and [SE-0111](https://github.com/apple/swift-evolution/blob/master/proposals/0111-remove-arg-label-type-significance.md).

File: test/ClangImporter/foreign_errors.swift (hunk 1 above)
Comment (slavapestov) [inline test/ClangImporter/foreign_errors.swift]: You mean it's a clang importer bug? I thought so...

**Comment by slavapestov** [comment]
@jrose-apple Yeah I'll definitely take a look at those. This is still a work in progress, it needs the staging flag before going in and in fact some of our libraries don't even build yet due to some other Sema bug I haven't figured out. So the language changes definitely are not final, I just wanted to see an approximation thereof.

File: test/ClangImporter/foreign_errors.swift (hunk 1 above)
Comment (jrose-apple) [inline test/ClangImporter/foreign_errors.swift]: We add a dummy argument when there's a naming conflict with another method, but we shouldn't when there's a conflict with a keyword.

**Comment by slavapestov** [comment]
Here is a new version where the corrected logic is conditionalized on the Swift 3 flag.

Note that unfortunately even without the fix, the behavior in master already diverged from Swift 3 in incompatible ways. I'm going to try to reconcile this.

**Comment by slavapestov** [comment]
FWIW the plan is for Swift 3 mode to support a superset of what the real Swift 3 supported. Also obviously I'm not going to try to simulate the old nonsense diagnostics and crashes...

**Comment by slavapestov** [comment]
@swift-ci Please smoke test

**Comment by slavapestov** [comment]
(Expecting a couple of failures in Compatibility/tuple_arguments)

**Comment by slavapestov** [comment]
Here are the source breaks I'm aware of so far with master vs Swift 3.0:

```let fn: (Int, Int) -> () = { t in _ = (t.0, t.1) }

let pong: (_ data : Any) -> () = {(data) in}
pong(data: "hello world")```

```/Users/slava/new/swift/test/Compatibility/tuple_arguments.swift:803:29: error: unexpected error produced: extra argument in call
  sTwo.genericFunction(3.0, 4.0) // FIXME
                            ^
/Users/slava/new/swift/test/Compatibility/tuple_arguments.swift:855:27: error: unexpected error produced: extra argument in call
  sTwo.genericFunction(a, b) // FIXME
                          ^
/Users/slava/new/swift/test/Compatibility/tuple_arguments.swift:1063:38: error: unexpected error produced: extra argument in call
  _ = GenericEnum<(Int, Int)>.one(3, 4) // FIXME
                                     ^```

**Comment by slavapestov** [comment]
@swift-ci Please smoke test

**Comment by slavapestov** [comment]
@swift-ci Please smoke test

**Comment by slavapestov** [comment]
@swift-ci Please smoke test

**Pull request merged**